product/gql_mutations.py

```diff
@@ -353,6 +356,70 @@ def async_mutate(cls, user, **data):
             ]
 
 
+class DuplicateProductMutation(OpenIMISMutation):
+    _mutation_module = "product"
+    _mutation_class = "DuplicateProductMutation"
+
+    class Input(ProductInputType):
+        code = graphene.String(required=True)
+        uuid = graphene.UUID(required=False)
+
+    @classmethod
+    def async_mutate(cls, user, **data):
+        try:
+            cls.do_mutate(
+                ProductConfig.gql_mutation_products_add_perms,
+                user,
+                **data,
+            )
+        except ValueError as exc:
+            return [
+                {
+                    "message": str(exc)
+                }
+            ]
+        except Exception as exc:
+            return [
+                {
+                    "message": _("product.mutation.failed_to_duplicate_product"),
+                    "detail": str(exc),
+                }
+            ]
+
+    @classmethod
+    def do_mutate(cls, perms, user, **data):
+        if type(user) is AnonymousUser or not user.id:
+            raise ValidationError(_("mutation.authentication_required"))
+        if not user.has_perms(perms):
+            raise PermissionDenied(_("unauthorized"))
+        current_uuid = data.pop("uuid") if "uuid" in data else None
+
+        data["audit_user_id"] = user.id_for_audit
+
+        duplicate_items = True if 'items' not in data else False
+        duplicate_services = True if 'services' not in data else False
+
+        new_product = create_or_update_product(user, data, is_duplicate=True)
+
+        if duplicate_items:
+            new_product_items = ProductItem.objects.filter(product=Product.objects.get(uuid=current_uuid,
+                                                           validity_to__isnull=True))
+            for item in new_product_items:
```

Comment:
are you stealing the item from the old product ? I don't see where you duplicate the item and services instance (create new item and then just copy the required fields)

Reply:
Yes, it is the only feasible way. ProductItem contains information about connection between Item and Product. I would have to query this either way and then insert the same info basically. This code copies those elements, old data does not get removed. See:
https://docs.djangoproject.com/en/4.1/topics/db/queries/#copying-model-instances

Reply:
@Thanks that make sense, I was afraid the "save_history" function was called upon `save()`